ClosureLifetimeFixup: Remove copy of borrowed move-only nonescaping captures when possible.

jckarter opened the pull request:
SILGen introduces a copy of the capture, because the semantics of escaping partial_apply's requires the closure to take ownership of the parameters. We don't know when a closure is strictly nonescaping or its final lifetime until ClosureLifetimeFixup runs, but that replaces the consume of the copy with a borrow of the copy normally, hoping later passes fix it up. We can't wait that long for move-only types, which can't be copied, so try to remove the copy up front when the copy lives long enough and has no interfering uses other than the partial_apply. rdar://110137169

jckarter committed:
ClosureLifetimeFixup: Remove copy of borrowed move-only nonescaping captures when possible.

SILGen introduces a copy of the capture, because the semantics of escaping partial_apply's
requires the closure to take ownership of the parameters. We don't know when a closure is
strictly nonescaping or its final lifetime until ClosureLifetimeFixup runs, but that replaces
the consume of the copy with a borrow of the copy normally, hoping later passes fix it up.
We can't wait that long for move-only types, which can't be copied, so try to remove the
copy up front when the copy lives long enough and has no interfering uses other than the
partial_apply. rdar://110137169

jckarter commented:
@swift-ci Please test

atrick reviewed (commented):
This is essentialy running a version of CopyForwarding inline with the existing partial apply rewrite. It's unfortunate that we need this kind of complexity inside such a narrow special purpose code path. It seems like this only applies to immutable captures though, so maybe it can all go away with opaque SIL values?

The new logic has lots of local state and doesn't need much from the enclosing context: cvt, closureOp, and newPA. It might be easy to factor.

>      if (closureLiveness.isWithinBoundary(origUse->getUser())) {
This doesn't look through address casts or projections. TransitiveAddressWalker does that. ~Sadly, TransitiveAddressWalker does not currently distinguish between transitive vs. leaf uses, and you only care about leaf uses.~ Also, it always sets AddressUseKind::PointerEscape for any other closure captures. I'm not sure if that will be too conservative. The plan is to fix TransitiveAddressWalker in both of these respects soon. In the meantime, you could just add a callback or two for your purpose.

Do you care if the original value is used or just if it is written? You could ask if the user `mayWriteToMemory` before checking if it's inside the temporary liveness.

Otherwise, the logic seems totally solid to me. It looks like the worst case is that we still fail to allow some patterns to pass move-only diagnostics. I like the use of tracing to track those down.

atrick commented:
> ~TransitiveAddressWalker does not currently distinguish between transitive vs. leaf uses, and you only care about leaf uses.~

That comment I made was wrong. TransitiveAddressWalker gives you the leaf uses, which is correct in your case.

jckarter commented:
@atrick I've updated the patch to skip over uses that `!mayWriteToMemory()` and added a test that we let the same borrow get captured multiple times in simultaneous nonescaping closures.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

gottesmm commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -289,7 +289,9 @@ void insertDestroyOfCapturedArguments(
 void insertDeallocOfCapturedArguments(PartialApplyInst *pai,
-                                      DominanceInfo *domInfo);
+    DominanceInfo *domInfo,
+    llvm::function_ref<bool(SILValue)> shouldInsertDestroy =
```
IIRC, you can do this more efficiently by passing in a nullptr as the default parameter for llvm::function_ref and just check if function_ref is nullptr before calling it. This means in the common case, we have a well predicated conditional instead of calling an indirect pointer.

My memory may be incorrect though.

jckarter replied:
`insertDestroyOfCapturedArguments` doesn't do that. Using `nullptr` also makes calling the function more mistake-prone since code has to check for null at every use site. A modern CPU is going to predict an indirect branch that always goes the same place about as well as a binary branch that does the same so I don't think a minor CPU resource optimization is really worth the added maintenance hazard.

gottesmm commented on test/SILOptimizer/moveonly_closure_lifetime_fixup_address_only_borrowed_captures.swift:
```diff
@@ -0,0 +1,64 @@
+// RUN: %target-swift-frontend -emit-sil %s | %FileCheck %s
+
+// When ClosureLifetimeFixup promotes a closure to the stack, we should
```
Can you add a SIL test case that shows this behavior? I would be fine with a single FileCheck test case that just shows the behavior in action.

gottesmm commented on lib/SILOptimizer/Mandatory/ClosureLifetimeFixup.cpp:
```diff
@@ -712,7 +714,186 @@ static SILValue tryRewriteToPartialApplyStack(
+    
+    OrigUnusedDuringClosureLifetimeWalker origUseWalker(closureLiveness,
+                                              origIsUnusedDuringClosureLifetime);
+    std::move(origUseWalker).walk(orig);
```
You may want for future proofing to add the check that we do not return AddressUseKind::Unknown. This is propagated back when you return false from the walk function. That being said, I do see that in the current iteration of the code origIsUnusedDuringClosureLifetime has the same effect. Your choice on what you want to do with this.

atrick replied:
If the result is anything other than `AddressUseKind::NonEscaping`, then the walker did not find all address uses. The problem is that other closure captures will report an escaping use even if the address only escapes immutably.

If this is too conservative, then my previous suggestion was to replace this code in the walker:
```
// TODO: Partial apply should be NonEscaping, but then we need to consider
    // the apply to be a use point.
    if (isa<PartialApplyInst>(user) || isa<AddressToPointerInst>(user)) {
      meet(AddressUseKind::PointerEscape);
      callVisitUse(op);
      continue;
    }
```
With a `visitCapture` override that returns `AddressUseKind`, which `OrigUnusedDuringClosureLifetimeWalker` could implement to return `NonEscaping` for immutable captures and `PointerEscape` for mutable captures.

atrick replied:
I tried to explain this better here: rdar://111160731 ([move-only] Handle pointer escapes in OrigUnusedDuringClosureLifetimeWalker)

gottesmm reviewed (approved):
LGTM. My main comments are:

1. Can you add a SIL test case that shows the behavior. Just something simple would be fine.
2. I included a few small improvements for performance. Feel free to think about them and use them if they make sense to you.

Add the SIL test case and the changes that I suggested if they work for you and lets get this committed!

gottesmm commented:
Also, with my suggestions, I am fine if you want to do a follow up so we can get this into 5.9.

jckarter commented:
@swift-ci Please test
